temurin-bin: init, semeru-bin: init, adoptopenjdk-bin: remove 13, 14, 17

taku0 opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

AdoptOpenJDK Hotspot is now Eclipse Temurin and AdoptOpenJDK OpenJ9 is IBM Semeru Runtime.

This PR adds `temurin-bin` and `semeru-bin`, and removes `adoptopenjdk-bin` package.

Temurin/Semeru 13, 14, and 15 is no longer supported and removed.

Supersedes #140116.

cc: @timothyklim, @edwtjo, @asbachb, @fpletz

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [X] x86_64-darwin
  - [ ] aarch64-darwin
- [X] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Tested with [a simple AWT and Swing sample app](https://github.com/taku0/JavaSample/tree/cdf6c68ad44a347953bc71ab45904c992eead37d).

timothyklim committed:
adoptopenjdk: add 17.0.1

taku0 committed:
temurin-bin, semeru-bin: init at 17.0.3, adoptopenjdk-bin: remove 13, 14, 17

AdoptOpenJDK Hotspot is now Eclipse Temurin and AdoptOpenJDK OpenJ9 is
now IBM Semeru Runtime.

Adds `temurin-bin` and `semeru-bin` packages.

AdoptOpenJDK 13, 14, and 15 is no longer supported.  AdoptOpenJDK 13 and
14 are removed.  AdoptOpenJDK 15 is left only for some packages
depending it.

SuperSandro2000 commented on pkgs/development/compilers/semeru-bin/jdk11-darwin.nix:
```diff
@@ -0,0 +1,7 @@
+let
+  sources = builtins.fromJSON (builtins.readFile ./sources.json);
```
Use lib.importJSON

SuperSandro2000 commented on pkgs/development/compilers/semeru-bin/jdk11-linux.nix:
```diff
@@ -0,0 +1,7 @@
+let
+  sources = builtins.fromJSON (builtins.readFile ./sources.json);
```
Use lib.importJSON

SuperSandro2000 commented on pkgs/development/compilers/temurin-bin/jdk-darwin-base.nix:
```diff
@@ -11,8 +15,8 @@ assert (stdenv.isDarwin && stdenv.isx86_64);
   name = if sourcePerArch.packageType == "jdk"
-    then "adoptopenjdk-${sourcePerArch.vmType}-bin-${sourcePerArch.${cpuName}.version}"
-    else "adoptopenjdk-${sourcePerArch.packageType}-${sourcePerArch.vmType}-bin-${sourcePerArch.${cpuName}.version}";
+         then "${name-prefix}-bin-${sourcePerArch.${cpuName}.version}"
```
The formatting before was correct.

SuperSandro2000 commented on pkgs/top-level/aliases.nix:
```diff
@@ -38,6 +38,20 @@ mapAliases ({
   accounts-qt = libsForQt5.accounts-qt; # added 2015-12-19
   adobeReader = adobe-reader; # added 2013-11-04
   adobe_flex_sdk = apache-flex-sdk; # added 2018-06-01
+  adoptopenjdk-hotspot-bin-17 = throw "AdoptOpenJDK is now Temurin. Use temurin-bin-17"; # added 2021-10-03
+  adoptopenjdk-jre-hotspot-bin-17 = throw "AdoptOpenJDK is now Temurin. Use temurin-jre-bin-17"; # added 2021-10-03
+  adoptopenjdk-hotspot-bin-15 = throw "AdoptOpenJDK 15 is no longer maintained"; # added 2021-10-03
+  adoptopenjdk-jre-hotspot-bin-15 = throw "AdoptOpenJDK 15 is no longer maintained"; # added 2021-10-03
+  adoptopenjdk-openj9-bin-15 = throw "AdoptOpenJDK 15 is no longer maintained"; # added 2021-10-03
+  adoptopenjdk-jre-openj9-bin-15 = throw "AdoptOpenJDK 15 is no longer maintained"; # added 2021-10-03
+  adoptopenjdk-hotspot-bin-14 = throw "AdoptOpenJDK 14 is no longer maintained"; # added 2021-10-03
+  adoptopenjdk-jre-hotspot-bin-14 = throw "AdoptOpenJDK 14 is no longer maintained"; # added 2021-10-03
+  adoptopenjdk-openj9-bin-14 = throw "AdoptOpenJDK 14 is no longer maintained"; # added 2021-10-03
+  adoptopenjdk-jre-openj9-bin-14 = throw "AdoptOpenJDK 14 is no longer maintained"; # added 2021-10-03
+  adoptopenjdk-hotspot-bin-13 = throw "AdoptOpenJDK 13 is no longer maintained"; # added 2021-10-03
+  adoptopenjdk-jre-hotspot-bin-13 = throw "AdoptOpenJDK 13 is no longer maintained"; # added 2021-10-03
+  adoptopenjdk-openj9-bin-13 = throw "AdoptOpenJDK 13 is no longer maintained"; # added 2021-10-03
+  adoptopenjdk-jre-openj9-bin-13 = throw "AdoptOpenJDK 13 is no longer maintained"; # added 2021-10-03
```
We don't add aliases for all version variants otherwise the aliases file would be to big.

SuperSandro2000 commented:
Please check with ``nix-env -qaP`` locally if everything evals.

taku0 commented:
@SuperSandro2000 
Addressed the review comments.

Temurin don't have version 15 while `openjdk16` requires it for bootstrapping.  So, I'm keeping AdoptOpenJDK until `openjdk16` is removed.

SuperSandro2000 commented on pkgs/development/compilers/temurin-bin/jdk-darwin-base.nix:
```diff
@@ -0,0 +1,69 @@
+    license = licenses.gpl2Classpath;
+    description = "${brand-name}, prebuilt OpenJDK binary";
+    platforms = [ "x86_64-darwin" ]; # some inherit jre.meta.platforms
+    maintainers = with lib.maintainers; [ taku0 ];
```
```suggestion
    maintainers = with maintainers; [ taku0 ];
```

SuperSandro2000 commented on pkgs/development/compilers/temurin-bin/jdk-linux-base.nix:
```diff
@@ -0,0 +1,126 @@
+  meta = with lib; {
+    license = licenses.gpl2Classpath;
+    description = "${brand-name}, prebuilt OpenJDK binary";
+    platforms = lib.mapAttrsToList (arch: _: arch + "-linux") sourcePerArch; # some inherit jre.meta.platforms
+    maintainers = with lib.maintainers; [ taku0 ];
```
```suggestion
    platforms = mapAttrsToList (arch: _: arch + "-linux") sourcePerArch; # some inherit jre.meta.platforms
    maintainers = with maintainers; [ taku0 ];
```

taku0 commented:
@SuperSandro2000 
Addressed the review comments.

FreeMonoid commented:
Anything that can be done here by an outsider to help move things along?

taku0 commented:
I will update it this weekends (or next weekends).

taku0 commented:
Resolved conflicts. Updated versions and tested again.

Moredread commented:
Mediathekview needs Java 17 for the latest version that also contains a security fix (upgrading log4j).

Is there a plan to backport this PR to 21.11?

pmenke-de commented on pkgs/development/compilers/temurin-bin/jdk-darwin-base.nix:
```diff
@@ -0,0 +1,67 @@
+  meta = with lib; {
+    license = licenses.gpl2Classpath;
+    description = "${brand-name}, prebuilt OpenJDK binary";
+    platforms = [ "x86_64-darwin" ]; # some inherit jre.meta.platforms
```
On an aarch64 Mac this causes unsupported-platform errors for me:
```
error: Package ‘temurin-bin-17.0.1’ in /nix/store/2amg850cxgs627bix28nm0m5zq4fv38k-source/pkgs/development/compilers/temurin-bin/jdk-darwin-base.nix:60 is not supported on ‘aarch64-darwin’, refusing to evaluate.
```

I tested around a bit and came to this solution: https://github.com/taku0/nixpkgs/compare/temurin-17...pmenke-de:pmenke-de_temurin-17

Maybe some of these adaptions could be used in this PR.
I'm not 100% sure on the `meta.platforms` end of things though, as I don't know how this works, when the same-named package is built from different sources on different platforms. Should `meta.platforms` list all theoretically supported platforms in this case, or should it only list the specific platform(s), this specific instance of the package runs on?

taku0 replied:
We should able to filter packages by platform without building it.  Since `version` or other attributes are evaluated lazily, the error is not thrown until the build time. Therefore, we should list actually supported platforms in `platforms` rather than relying on errors on the build time.

Because I don't have M1 Macs, I'm hesitating to add `aarch64-darwin` to `platforms`.

pmenke-de replied:
> We should able to filter packages by platform without building it. Since version or other attributes are evaluated lazily, the error is not thrown until the build time. Therefore, we should list actually supported platforms in platforms rather than relying on errors on the build time.

Ok. That's good to know. Thank you.

> Because I don't have M1 Macs, I'm hesitating to add `aarch64-darwin` to `platforms`.

Understandable. I do have a M1 Mac though and thus would appreciate, if we could include `aarch64-darwin` as supported platform. If I could somehow help this effort - e.g. by testing or sth. please let me know.
So far I can at least say, that it builds and runs on my machine :-)

In my above mentioned commit (https://github.com/taku0/nixpkgs/compare/temurin-17...pmenke-de:pmenke-de_temurin-17) - that I changed/updated in the meantime - I also addressed one more thing:

This line in [jdk-linux-base.nix](https://github.com/NixOS/nixpkgs/pull/140364/files#diff-8b64e051aa30db104a76317d4ca010753c1415eb83d5b1e75865b0ba1ac8def7R119) generates some unexpected platform entries, from non-platform like keys in the sources.json:
```nix
platforms = lib.mapAttrsToList (arch: _: arch + "-linux") sourcePerArch
```

```nix
nix-repl> outputs.legacyPackages.x86_64-linux.temurin-bin-11.meta.platforms
[ "aarch64-linux" "armv6l-linux" "armv7l-linux" "packageType-linux" "vmType-linux" "x86_64-linux" ]
```

Maybe sth. like this would be better? - just looking at the expression output; I'm not experienced enough to tell whether this is a "well-written" nix-expression.
```nix
platforms = builtins.map (arch: arch + "-linux") (builtins.filter ( arch: builtins.elem arch (builtins.attrNames lib.systems.parse.cpuTypes) ) (builtins.attrNames sourcePerArch));
```
```nix
nix-repl> outputs.legacyPackages.x86_64-linux.temurin-bin-11.meta.platforms
[ "aarch64-linux" "armv6l-linux" "armv7l-linux" "x86_64-linux" ]
```

taku0 replied:
@pmenke-de I have incorporated your idea. Thank you. Could you try it on M1 Mac?

pmenke-de replied:
@taku0 I tested it on my machine. Works perfectly 👍 
Thanks for implementing it.

taku0 commented:
Now using `pname` and `version` rather than `name`.

taku0 commented:
Fixed ofborg errors.

andreoss commented:
@taku0 When is it going to be merged?

taku0 commented:
@andreoss We need a review and approval.

Lassulus commented:
needs a rebase

mschuwalow commented:
@taku0 are you still working on this?

taku0 commented:
I will update this on the weekend. Maybe it is better to focus on temurin and semeru.

taku0 commented:
Updated.

mschuwalow commented on pkgs/top-level/all-packages.nix:
```diff
@@ -12388,113 +12388,49 @@ with pkgs;
-  adoptopenjdk-jre-openj9-bin-8 = if stdenv.isLinux
-    then callPackage adoptopenjdk-bin-8-packages-linux.jre-openj9 {}
-    else callPackage adoptopenjdk-bin-8-packages-darwin.jre-openj9 {};
+  temurin-bin-17 = javaPackages.compiler.temurin-bin.jdk-17;
+  temurin-bin-16 = javaPackages.compiler.temurin-bin.jdk-16;
+  # AdoptOpenJDK has JRE 16 but Temurin does't.
+  temurin-jre-bin-16 = javaPackages.compiler.temurin-bin-16;
+  temurin-bin-11 = javaPackages.compiler.temurin-bin.jdk-11;
+  temurin-jre-bin-11 = javaPackages.compiler.temurin-bin.jre-11;
+  temurin-bin-8 = javaPackages.compiler.temurin-bin.jdk-8;
+  temurin-jre-bin-8 = javaPackages.compiler.temurin-bin.jre-8;
+
+  temurin-bin = temurin-bin-17;
+  temurin-jre-bin = temurin-bin;
+
+  semeru-bin-17 = javaPackages.compiler.semeru-bin.jdk-17;
+  semeru-jre-bin-17 = javaPackages.compiler.semeru-bin.jre-17;
+  semeru-bin-16 = javaPackages.compiler.semeru-bin.jdk-16;
+  semeru-jre-bin-16 = javaPackages.compiler.semeru-bin.jre-16;
+  semeru-bin-11 = javaPackages.compiler.semeru-bin.jdk-11;
+  semeru-jre-bin-11 = javaPackages.compiler.semeru-bin.jre-11;
+  semeru-bin-8 = javaPackages.compiler.semeru-bin.jdk-8;
+  semeru-jre-bin-8 = javaPackages.compiler.semeru-bin.jre-8;
+
+  semeru-bin = semeru-bin-11;
+  semeru-jre-bin = semeru-bin;
```
```suggestion
  temurin-bin-17 = javaPackages.compiler.temurin-bin.jdk-17;
  temurin-jre-bin-17 = javaPackages.compiler.temurin-bin.jre-17;
  temurin-bin-16 = javaPackages.compiler.temurin-bin.jdk-16;
  temurin-jre-bin-16 = javaPackages.compiler.temurin-bin.jre-16;
  temurin-bin-11 = javaPackages.compiler.temurin-bin.jdk-11;
  temurin-jre-bin-11 = javaPackages.compiler.temurin-bin.jre-11;
  temurin-bin-8 = javaPackages.compiler.temurin-bin.jdk-8;
  temurin-jre-bin-8 = javaPackages.compiler.temurin-bin.jre-8;

  temurin-bin = temurin-bin-17;
  temurin-jre-bin = temurin-jre-bin-17;

  semeru-bin-17 = javaPackages.compiler.semeru-bin.jdk-17;
  semeru-jre-bin-17 = javaPackages.compiler.semeru-bin.jre-17;
  semeru-bin-16 = javaPackages.compiler.semeru-bin.jdk-16;
  semeru-jre-bin-16 = javaPackages.compiler.semeru-bin.jre-16;
  semeru-bin-11 = javaPackages.compiler.semeru-bin.jdk-11;
  semeru-jre-bin-11 = javaPackages.compiler.semeru-bin.jre-11;
  semeru-bin-8 = javaPackages.compiler.semeru-bin.jdk-8;
  semeru-jre-bin-8 = javaPackages.compiler.semeru-bin.jre-8;

  semeru-bin = semeru-bin-17;
  semeru-jre-bin = semeru-jre-bin-17;
```

Why not expose the full matrix of versions? They seem to work fine on my machine.

taku0 replied:
Those JREs were not available when this PR was written.  I have incorporated your suggestions.

mschuwalow commented on pkgs/development/compilers/temurin-bin/jdk-darwin.nix:
```diff
@@ -0,0 +1,14 @@
+  common = opts: callPackage (import ./jdk-darwin-base.nix opts) {};
+in
+{
+  jdk-8 = common { sourcePerArch = sources.jdk.openjdk8; };
+  jre-8 = common { sourcePerArch = sources.jre.openjdk8; };
+  jdk-11 = common { sourcePerArch = sources.jdk.openjdk11; };
+  jre-11 = common { sourcePerArch = sources.jre.openjdk11; };
+  jdk-16 = common { sourcePerArch = sources.jdk.openjdk16; };
+  jdk-17 = common { sourcePerArch = sources.jdk.openjdk17; };
```
```suggestion
  jdk-8 = common { sourcePerArch = sources.jdk.openjdk8; };
  jre-8 = common { sourcePerArch = sources.jre.openjdk8; };
  jdk-11 = common { sourcePerArch = sources.jdk.openjdk11; };
  jre-11 = common { sourcePerArch = sources.jre.openjdk11; };
  jdk-16 = common { sourcePerArch = sources.jdk.openjdk16; };
  jre-16 = common { sourcePerArch = sources.jre.openjdk16; };
  jdk-17 = common { sourcePerArch = sources.jdk.openjdk17; };
  jre-17 = common { sourcePerArch = sources.jre.openjdk17; };
```

mschuwalow commented on pkgs/development/compilers/temurin-bin/jdk-linux.nix:
```diff
@@ -0,0 +1,14 @@
+  common = opts: callPackage (import ./jdk-linux-base.nix opts) {};
+in
+{
+  jdk-8 = common { sourcePerArch = sources.jdk.openjdk8; };
+  jre-8 = common { sourcePerArch = sources.jre.openjdk8; };
+  jdk-11 = common { sourcePerArch = sources.jdk.openjdk11; };
+  jre-11 = common { sourcePerArch = sources.jre.openjdk11; };
+  jdk-16 = common { sourcePerArch = sources.jdk.openjdk16; };
+  jdk-17 = common { sourcePerArch = sources.jdk.openjdk17; };
```
```suggestion
  jdk-8 = common { sourcePerArch = sources.jdk.openjdk8; };
  jre-8 = common { sourcePerArch = sources.jre.openjdk8; };
  jdk-11 = common { sourcePerArch = sources.jdk.openjdk11; };
  jre-11 = common { sourcePerArch = sources.jre.openjdk11; };
  jdk-16 = common { sourcePerArch = sources.jdk.openjdk16; };
  jre-16 = common { sourcePerArch = sources.jre.openjdk16; };
  jdk-17 = common { sourcePerArch = sources.jdk.openjdk17; };
  jre-17 = common { sourcePerArch = sources.jre.openjdk17; };
```

SuperSandro2000 commented:
Could we easily change this to have less boilerplate code?

mschuwalow commented on pkgs/development/compilers/temurin-bin/jdk-darwin.nix:
```diff
@@ -0,0 +1,16 @@
+  jdk-11 = common { sourcePerArch = sources.jdk.openjdk11; };
+  jre-11 = common { sourcePerArch = sources.jre.openjdk11; };
+  jdk-16 = common { sourcePerArch = sources.jdk.openjdk16; };
+  jre-16 = common { sourcePerArch = sources.jre.openjdk16; };
```
```suggestion
```

mschuwalow commented on pkgs/top-level/all-packages.nix:
```diff
@@ -12388,113 +12388,49 @@ with pkgs;
+  temurin-bin-17 = javaPackages.compiler.temurin-bin.jdk-17;
+  temurin-jre-bin-17 = javaPackages.compiler.temurin-bin.jre-17;
+  temurin-bin-16 = javaPackages.compiler.temurin-bin.jdk-16;
+  temurin-jre-bin-16 = javaPackages.compiler.temurin-bin.jre-16;
```
```suggestion
```

mschuwalow reviewed (commented):
sorry for my last comment, seems like the jre one is still missing for openjdk16. All others should be fine though

taku0 commented:
Addressed the comments.

NickCao commented:
ping @SuperSandro2000 in case you forgot this one 😉

taku0 commented:
Updated to the latest version.

fpletz commented:
Sorry I've not looked at this PR sooner. I've recently dropped maintainership for AdoptOpenJDK in nixpkgs but I've looked at the changes now and they look good to me. Since there are evidently users of this, let's merge this now. Thanks!
